### PR TITLE
Set close-on-exec for pipe writer in console mode

### PR DESCRIPTION
**Pull request by axw**
The pipe isn't used by subprocesses in console mode,
so close it. This avoids the possibility of the FD
being leaked into nested subprocesses, causing Ninja
to hang like in #965.

### CODE REVIEW

File: src/subprocess-posix.cc [hunk 1]
```diff
@@ -55,8 +55,6 @@ bool Subprocess::Start(SubprocessSet* set, const string& command) {
     Fatal("fork: %s", strerror(errno));
 
   if (pid_ == 0) {
-    close(output_pipe[0]);
```
Comment (pcc) [inline src/subprocess-posix.cc]: Why is this needed? This is the parent's side of the pipe so we ought to close it, right?

File: src/subprocess-posix.cc [hunk 2]
```diff
@@ -90,9 +88,12 @@ bool Subprocess::Start(SubprocessSet* set, const string& command) {
         // Now can use stderr for errors.
         error_pipe = 2;
         close(output_pipe[1]);
+      } else {
+        // In the console case, the pipe is not needed by the child
+        // process at all. Close it in the child, but keep it open
+        // in the parent in case execl fails.
+        SetCloseOnExec(output_pipe[1]);
```
Comment (pcc) [inline src/subprocess-posix.cc]: I think this will mean that Ninja will be unable to start tasks in parallel while the console pool task is running, as it will be stuck in the `waitpid` call in `Subprocess::Finish`. Maybe we should be watching for SIGCHLD from pool=console child processes instead of waiting on the pipe.

File: src/subprocess-posix.cc (hunk 1 above)
Comment (axw) [inline src/subprocess-posix.cc]: This is a drive-by; it's already set O_CLOEXEC, immediately before the fork.

File: src/subprocess-posix.cc (hunk 2 above)
Comment (axw) [inline src/subprocess-posix.cc]: Ah, I think I see. I'll paraphrase below, please let me know if I misunderstand.

DoWork/OnPipeReady is using the pipe to determine whether or not the child is done. If there were a console pool task, it would immediately go do Done, and then Finish would block until it's really done. While blocking, other tasks' buffers would fill up, potentially blocking them from doing work.

For *only* console pool tasks, DoWork could instead use waitpid. That would have the same effect as ppoll on a non-console task that never sends anything to stdout (i.e. they'd both wait until the subprocess dies).

**Comment by axw** [comment]
@pcc Please see latest commit. It should better handle parallel tasks now. See commit message for a description, if necessary.

**Comment by maximuska** [comment]
- Couldn't understand what makes _console_ case special with respect to
process termination detection to trigger the issue. If we understand the
problem, is it possible to create a small repro, like something running
some bash one-line script?
- @andrew: wouldn't setting sigchld to wakeup ppoll and doing a loop of
waitpid(-1, &stat, WNOHANG) in DoWork() on EINTR be a simpler change?

**Comment by axw** [comment]
@maximuska You're right. There's nothing console-specific about it, and my change doesn't cater for non-console. Repro build.ninja:

```
rule test
    command = sleep 10s&
    pool = console # or without
build test: test
default test
```

I will investigate your suggestion.
